SA3/qa_web/views.py

```diff
@@ -102,6 +102,48 @@ def answers(request, id_):
     return render(request, 'qa_web/answerspage.html', {'currentQuestion': q, 'answers': q_answers, 'bestAnswer': q_best_answer})
 
 
+def vote(request):
+    """Receives Ajax queries to vote on Posts, updates the corresponding Post and returns the new score"""
+    if request.method == 'POST' and request.user.is_authenticated:
+        vote_direction, button_id, post_type = request.POST['button'].split('_')
+        vote_direction = vote_direction == 'upvote'
+        post_subclass = Questions if post_type == 'question' else Answers if post_type == 'answer' else Comments
+        post = get_object_or_404(post_subclass, pk=int(button_id))
+        foreign_key_args = {post_type:post} # Workaround to provide the correct kwarg depending on post_type
+        if request.user not in post.voters.all():
+            # Add user's vote
+            Vote.objects.create(user=request.user, positive=vote_direction,
+                                    **foreign_key_args) # Bind the post to the appropriate FK
+            if vote_direction:
+                post.upvotes = post.upvotes + 1
+            else:
+                post.downvotes = post.downvotes + 1
+        else:
+            # Allow user to modify their vote
+            last_vote = Vote.objects.get(user=request.user, **foreign_key_args)
+            if last_vote.positive == vote_direction:
+                # Vote cancelled, user pressed same button as original vote
+                last_vote.delete()
+                if vote_direction:
+                    post.upvotes = post.upvotes - 1
+                else:
+                    post.downvotes = post.downvotes - 1
+            else:
+                # Vote changed direction
+                last_vote.positive = vote_direction
+                last_vote.save()
+                if vote_direction:
+                    post.upvotes = post.upvotes + 1
+                    post.downvotes = post.downvotes - 1
+                else:
+                    post.downvotes = post.downvotes + 1
+                    post.upvotes = post.upvotes - 1
+        post.save()
+        return JsonResponse({'new_score' : post.score, 'id' : 'score_{}_{}'.format(post.id, post_type)})
```

Comment:
Nice. Sending the data back to the label that indicates the score after successful response to ajax request.